src/otaclient/boot_control/_base.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2022 TIER IV, INC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Common base class for boot controllers to reduce code duplication."""
+
+from __future__ import annotations
+
+import logging
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import NoReturn
+
+from otaclient import errors as ota_errors
+from otaclient._types import OTAStatus
+
+from ._ota_status_control import OTAStatusFilesControl
+from ._slot_mnt_helper import SlotMountHelper
+
+logger = logging.getLogger(__name__)
+
+
+class BootControllerBase(ABC):
+    """Abstract base class for boot controllers implementing common patterns.
+
+    This class implements the Template Method pattern to reduce code duplication
+    across different boot controller implementations (GRUB, Jetson UEFI, Jetson cboot, RPI).
+    Note: This class provides concrete implementations of BootControllerProtocol methods.
+    """
+
+    # These must be initialized by subclasses in __init__
+    _mp_control: SlotMountHelper
+    _ota_status_control: OTAStatusFilesControl
+
+    # ====== Abstract properties that subclasses must implement ======
+
+    @property
+    @abstractmethod
+    def bootloader_type(self) -> str:
+        """The type of bootloader. Must be implemented by subclasses."""
+        ...
```

Comment:
In `otaclient.configs._ecu_info` module, we have BootloaderType enum, let's use that here.